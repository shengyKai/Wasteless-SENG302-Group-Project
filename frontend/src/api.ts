/*
 * Created on Wed Feb 10 2021
 *
 * The Unlicense
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or distribute
 * this software, either in source code form or as a compiled binary, for any
 * purpose, commercial or non-commercial, and by any means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors of this
 * software dedicate any and all copyright interest in the software to the public
 * domain. We make this dedication for the benefit of the public at large and to
 * the detriment of our heirs and successors. We intend this dedication to be an
 * overt act of relinquishment in perpetuity of all present and future rights to
 * this software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

/**
 * Declare all available services here
 */
import axios from 'axios';
const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 2000,
  withCredentials: true,
});

type MaybeError<T> = T | string;

export type User = {
  id: number,
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  created?: string,
  role?: "user" | "globalApplicationAdmin" | "defaultGlobalApplicationAdmin",
  businessesAdministered?: Business[],
};

export type Location = {
  streetNumber?: string,
  streetName?: string,
  city?: string,
  region?: string,
  country: string,
  postcode?: string,
};

export type CreateUser = {
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  password: string,
};

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';

export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];

export type Business = {
  id: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
  created?: string,
};

export type CreateBusiness = {
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
};

export type Image = {
  id: number,
  filename: string,
  thumbnailFilename: string,
};

export type Product = {
  id: string,
  name: string,
  description?: string,
  manufacturer?: string,
  recommendedRetailPrice?: number,
  created?: string,
  images: Image[],
};

export type CreateProduct = Omit<Product, 'created' | 'images'>;

function isLocation(obj: any): obj is Location {
  if (obj === null || typeof obj !== 'object') return false;
  if (obj.streetNumber !== undefined && typeof obj.streetNumber !== 'string') return false;
  if (obj.streetName !== undefined && typeof obj.streetName !== 'string') return false;
  if (obj.city !== undefined && typeof obj.city !== 'string') return false;
  if (obj.region !== undefined && typeof obj.region !== 'string') return false;
  if (typeof obj.country !== 'string') return false;
  if (obj.postcode !== undefined && typeof obj.postcode !== 'string') return false;

  return true;
}

function isUser(obj: any): obj is User {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (typeof obj.firstName !== 'string') return false;
  if (typeof obj.lastName !== 'string') return false;
  if (obj.middleName !== undefined && typeof obj.middleName !== 'string') return false;
  if (obj.nickname !== undefined && typeof obj.nickname !== 'string') return false;
  if (obj.bio !== undefined && typeof obj.bio !== 'string') return false;
  if (typeof obj.email !== 'string') return false;
  if (typeof obj.dateOfBirth !== 'string') return false;
  if (obj.phoneNumber !== undefined && typeof obj.phoneNumber !== 'string') return false;
  if (!isLocation(obj.homeAddress)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;
  if (obj.role !== undefined && !['user', 'globalApplicationAdmin', 'defaultGlobalApplicationAdmin'].includes(obj.role))
    return false;
  if (obj.businessesAdministered !== undefined && !isBusinessArray(obj.businessesAdministered)) return false;

  return true;
}

function isBusiness(obj: any): obj is Business {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (obj.administrators !== undefined && !isUserArray(obj.administrators)) return false;
  if (typeof obj.name !== 'string') return false;
  if (obj.description !== undefined && typeof obj.description !== 'string') return false;
  if (typeof obj.address !== 'object') return false;
  if (!BUSINESS_TYPES.includes(obj.businessType)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;

  return true;
}

function isNumberArray(obj: any): obj is number[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (typeof elem !== 'number') return false;
  }
  return true;
}

function isUserArray(obj: any): obj is User[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isUser(elem)) return false;
  }
  return true;
}

function isBusinessArray(obj: any): obj is Business[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isBusiness(elem)) return false;
  }
  return true;
}

type OrderBy = 'userId' | 'relevance' | 'firstName' | 'middleName' | 'lastName' | 'nickname' | 'email' | 'address';

/**
 * Sends a search query to the backend.
 *
 * @param query Query string to search for
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of user infos for the current page or an error message
 */
export async function search(query: string, pageIndex: number, resultsPerPage: number, orderBy: OrderBy, reverse: boolean): Promise<MaybeError<User[]>> {
  let response;
  try {
    response = await instance.get('/users/search', {
      params: {
        searchQuery: query,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUserArray(response.data)) {
    return 'Response is not user array';
  }

  return response.data;
}

/**
 * Sends a query for the number of search results for a given query to the backend.
 *
 * @param query Query string to search for
 * @returns Number of search results or an error message
 */
export async function getSearchCount(query: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.get('/users/search/count', {
      params: {
        searchQuery: query,
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (typeof response.data?.count !== 'number') {
    return 'Response is not number';
  }

  return response.data.count;
}

/**
 * Queries the backend for a specific user by their id.
 *
 * @param id User id, if ommitted then fetch the logged in user's info
 * @returns User info for the given id or an error message
 */
export async function getUser(id: number): Promise<MaybeError<User>> {
  let response;
  try {
    response = await instance.get('/users/' + id);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUser(response.data)) {
    return 'Response is not user';
  }

  return response.data;
}

/**
 * Logs in to the given user account by setting the authentication cookie.
 *
 * @param email User email
 * @param password User password
 * @returns The now logged in user ID if operation is successful, otherwise a string error.
 */
export async function login(email: string, password: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.post('/login', {
      email: email,
      password: password,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid credentials';
    return `Request failed: ${status}`;
  }
  let id = response.data.userId;
  if (typeof id !== 'number') return 'Invalid response';

  return id;
}

/**
 * Creates a user with the given properties.
 * Note: This doesn't automatically log the user in.
 *
 * @param user Initial user properties
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createUser(user: CreateUser): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/users', user);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid create user request';
    if (status === 409) return 'Email in use';
    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Makes the given user a GAA, if the current user is a DGAA.
 *
 * @param userId User to make GAA
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function makeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/makeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Revokes the given user's admin rights
 *
 * @param userId User to revoke permissions
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function revokeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/revokeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Add a product to a businesses catalogue.
 *
 * @param businessId The business to add the product to
 * @param product The properties to create a product with
 * @return undefined if operation is successful, otherwise a string error
 */
export async function createProduct(businessId: number, product: CreateProduct): Promise<MaybeError<undefined>> {
  try {
    await  instance.post(`/businesses/${businessId}/products`, product);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';

    if (status === 400) {
      // TODO Not sure exactly how the backend is going to communicate with us that the product code
      // is unavailable.

      // eslint-disable-next-line no-constant-condition
      if (false) {
        return 'Product code unavailable';
      }

      return 'Invalid parameters';
    }

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!isBusiness(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}