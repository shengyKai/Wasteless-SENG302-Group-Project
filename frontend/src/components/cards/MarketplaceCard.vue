<template>
  <v-card min-height="250px" class="d-flex flex-column">

    <v-card-title class="my-n1 title">
      <div class="mr-1">
        {{ content.title }}
      </div>
      <v-chip v-if="showSection" color="primary" small outlined>
        {{ section }}
      </v-chip>
    </v-card-title>
    <v-card-text class="my-n2 flex-grow-1 d-flex flex-column justify-space-between">
      <div>
        <strong v-if="location">
          {{ locationString }}
          <br>
        </strong>
        <div class="d-flex justify-space-between flex-wrap">
          <span class="mr-1">
            <em v-if="creator">By {{ creator.firstName }} {{ creator.lastName }}</em>
          </span>
          <span v-if="content.created !== undefined">
            {{ dateString }}
          </span>
        </div>
        {{ content.description }}
      </div>
      <div class="v-flex mx-n1">
        <v-chip
          v-for="keyword in content.keywords"
          :key="keyword.id"
          small
          color="primary"
          class="mx-1 my-1"
        >
          {{ keyword.name }}
        </v-chip>
      </div>
    </v-card-text>
    <v-divider/>
    <v-card-actions v-if="isCardOwnerOrDGAA && showActions">
      <v-icon ref="deleteButton"
              color="primary"
              @click.stop="deleteCardDialog = true"
      >
        mdi-trash-can
      </v-icon>
      <v-dialog
        v-model="deleteCardDialog"
        max-width="300px"
      >
        <v-card>
          <v-card-title>
            Are you sure?
          </v-card-title>
          <v-card-text>
            Deleting "{{ content.title }}" will remove the card listing from the marketplace
          </v-card-text>
          <v-card-actions>
            <v-spacer/>
            <v-btn
              color="primary"
              text
              @click="deleteCard(content.id); deleteCardDialog = false;"
            >
              Delete
            </v-btn>
            <v-btn
              color="primary"
              text
              @click="deleteCardDialog = false"
            >
              Cancel
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-dialog>
      <v-btn
        @click="editCardDialog=true">
        Temporary edit button
      </v-btn>
      <template v-if="editCardDialog">
        <MarketplaceCardForm :user="$store.state.user" :previousCard="content" @closeDialog="editCardDialog=false"/>
      </template>
    </v-card-actions>
  </v-card>
</template>

<script>
import { formatDate, SECTION_NAMES } from '@/utils';
import { deleteMarketplaceCard } from '../../api/internal.ts';
import MarketplaceCardForm from '../marketplace/MarketplaceCardForm.vue';

export default {
  name: "MarketplaceCard",
  data () {
    return {
      deleteCardDialog: false,
      editCardDialog: false
    };
  },
  components: {
    MarketplaceCardForm,
  },
  props: {
    content: {
      id: Number,
      title: String,
      description: String,
      creator: Object,
      created: String,
      lastRenewed: String,
      keywords: Array,
    },
    showActions: {
      type: Boolean,
      default: false,
    },
    showSection: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    creator() {
      return this.content.creator;
    },
    location() {
      return this.creator?.homeAddress;
    },
    locationString() {
      if (this.location.district !== undefined && this.location.city !== undefined) {
        return `From ${this.location.district}, ${this.location.city}`;
      } else if (this.location.city !== undefined) {
        return `From ${this.location.city}, ${this.location.country}`;
      } else {
        return `From ${this.location.country}`;
      }
    },
    dateString() {
      let dateString = "Posted " + formatDate(this.content.created);
      if (this.content.lastRenewed !== this.content.created) {
        dateString += ", Renewed " + formatDate(this.content.lastRenewed);
      }
      return dateString;
    },
    // To ensure only the card owner, DGAA or GAA is able to execute an action relating to the marketplace card
    isCardOwnerOrDGAA() {
      return (this.$store.state.user.id === this.content.creator.id)
            || (this.$store.getters.role === "defaultGlobalApplicationAdmin")
            || (this.$store.getters.role === "globalApplicationAdmin");
    },
    section() {
      return SECTION_NAMES[this.content.section];
    },
  },

  methods: {
    formatDate,
    /**
     * Deletes the selected marketplace card and emits the response to the parent, Marketplace
     */
    async deleteCard(cardId) {
      let response = await deleteMarketplaceCard(cardId);
      this.$emit("delete-card", response);
    }
  }
};
</script>


<style scoped>
.title {
  line-height: 1.25;
  word-break: break-word;
}
</style>