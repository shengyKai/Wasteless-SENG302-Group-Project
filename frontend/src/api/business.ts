import {is} from 'typescript-is';
import {Location, MaybeError, SearchResults, instance} from "@/api/internal";
import {User} from "@/api/user";
import { Image } from '@/api/images';

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';
export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];
type BusinessOrderBy = 'created' | 'name' | 'location' | 'businessType';

export type Business = {
  id: number,
  primaryAdministratorId: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
  created?: string,
  images?: Image[],
};

export type CreateBusiness = {
  primaryAdministratorId: number,
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
};

export type ModifyBusiness = {
  primaryAdministratorId: number,
  name: string,
  description?: string,
  address: Location,
  businessType: BusinessType,
  updateProductCountry: boolean
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';

    return error.response.data.message;
  }

  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!is<Business>(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}

/**
 * Sends a business search query to the backend. Can specify query, business type or both.
 *
 * @param query Query string to search for
 * @param businessType Business type to filter search results by
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of business infos for the current page or an error message
 */
export async function searchBusinesses(query: string | undefined, businessType: BusinessType | undefined, pageIndex: number, resultsPerPage: number, orderBy: BusinessOrderBy, reverse: boolean): Promise<MaybeError<SearchResults<Business>>> {
  let response;
  try {
    response = await instance.get('/businesses/search', {
      params: {
        searchQuery: query,
        businessType: businessType,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid search query: ' + error.response?.data.message;
    if (status === 401) return 'You have been logged out. Please login again and retry';
    return `Request failed: ${status}`;
  }

  if (!is<SearchResults<Business>>(response.data)) {
    return 'Response is not business array';
  }

  return response.data;
}

/**
 * Makes the provided user an administrator of the provided business.
 *
 * @param businessId Business id to add an administrator to
 * @param userId User id to add to the provided business
 * @returns undefined if operation is successful, otherwise a string error message
 */
export async function makeBusinessAdmin(businessId: number, userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}/makeAdministrator`, {
      userId,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'User doesn\'t exist or is already an admin';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Current user cannot perform this action';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Removes the administrator status of a user from a business
 *
 * @param businessId Business id to remove an administrator from
 * @param userId User id to remove from the provided business
 * @returns undefined if operation is successful, otherwise a string error message
 */
export async function removeBusinessAdmin(businessId: number, userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}/removeAdministrator`, {
      userId,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'User doesn\'t exist or is not an admin';
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Current user cannot perform this action';
    if (status === 406) return 'The new business admin should be at least 16 years old';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Modifies a business given a business id and an updated business object
 * @param businessId Id of the business which is to be updated
 * @param business Business object with all the fields to be updated.
 */
export async function modifyBusiness(businessId: number, business: ModifyBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.put(`/businesses/${businessId}`, business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid details entered: ' + error.response?.data.message;
    if (status === 401) return 'You have been logged out. Please login again and retry';
    if (status === 403) return 'Invalid authorization for modifying this business';
    if (status === 406) return 'Business does not exist';

    return 'Request failed: ' + error.response?.data.message;
  }
  return undefined;
}