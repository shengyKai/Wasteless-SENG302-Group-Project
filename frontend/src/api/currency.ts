import { MaybeError } from "@/api/internal";

/**
 * Currency information for a specific country.
 */
export type Currency = {
  code: string,
  name: string,
  symbol: string
};

type CurrencyOrError = Currency | { errorMessage: string }

/**
 * An object which only has the attribute 'currencies', which is a list of Currency objects. The API response is expected
 * to contain an array of objects of this type
 */
type CurrenciesContainer = {
  currencies: Currency[];
};

/**
 * Determines if the given object is an instance of Currency.
 * @param obj The object to be checked.
 * @returns True if the object has the attributes of a Currency object, false otherwise.
 */
function isCurrency(obj: any): obj is Currency {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.code !== 'string') return false;
  if (typeof obj.name !== 'string') return false;
  if (typeof obj.symbol !== 'string') return false;
  return true;
}

/**
 * Determines if the given object is an array of object of type Currency.
 * @param obj The object to be checked.
 * @returns True if the object is an array where all elements are Currency objects, false otherwise.
 */
function isCurrencyArray(obj: any): obj is Currency[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isCurrency(elem)) return false;
  }
  return true;
}

/**
 * Check that the body of the API response has the expected format. It should be an array of length 1, where they
 * entry in the array is a JSON with a "currencies" field, and that field contains an array of currency objects.
 * @param response The body of the response received from the API.
 * @returns True if the response is in the expected format, false otherwise.
 */
function currencyResponseHasExpectedFormat(response: any): response is [CurrenciesContainer] {
  if (!Array.isArray(response)) return false;
  if (response.length !== 1) return false;
  const country = response[0];
  if (!isCurrencyArray(country.currencies)) return false;
  return true;
}

/**
 * Make a request to the RESTCounties API to find the currency associated with the given country name.
 * If the request is successful the currency from the API will be returned. If it is unsuccessful then
 * a default currency object with blank code, name and symbol fields will be returned.
 * @param country The name of a country to use in the API request for the currency.
 * @returns A list containing the Currency of the country(defaultCurrency if any errors occur) at index 0,
 * and an error message at index 1 if any errors occur.
 */
export async function currencyFromCountry(country: string): Promise<CurrencyOrError> {

  const response = await queryCurrencyAPI(country);

  if (typeof response === 'string') {
    console.warn(response);
    return { errorMessage: response };
  }

  const currency = await getCurrencyFromAPIResponse(response);

  if (typeof currency === 'string') {
    console.warn(currency);
    return { errorMessage: currency };
  }

  return currency;
}

/**
 * This method takes a string with the name of a country and queries the RESTCountries API to find the
 * currency associated with that country. The API's response will be returned if a response with status
 * code 200 is received, otherwise an error message will be returned.
 * @param country The name of the country to query the API for.
 * @return the response received from the RESTCounties API or a string error message.
 */
export async function queryCurrencyAPI(country: string): Promise<MaybeError<Response>> {

  const queryUrl = `https://restcountries.eu/rest/v2/name/${country}?fullText=true&fields=currencies`;

  const response = await fetch(queryUrl)
    .catch(error => {
      return `Failed to reach ${queryUrl}`;
    });
  if (typeof response === 'string') {
    return response;
  }
  if (response.status === 404) {
    return `No country with name ${country} was found`;
  }
  if (response.status !== 200) {
    return `Request failed: ${response.status}`;
  }
  return response;
}

/**
 * This method checks the format of the API response and extracts a currency object from the JSON body
 * if the response format is correct. If it is not correct then an error message is returned.
 * @param response A currency object extracted from the response body or an error message.
 */
export async function getCurrencyFromAPIResponse(response: Response): Promise<MaybeError<Currency>> {
  const responseBody = await response.json();

  if (!currencyResponseHasExpectedFormat(responseBody)) {
    return 'API response was not in readable format';
  }

  return responseBody[0].currencies[0];
}