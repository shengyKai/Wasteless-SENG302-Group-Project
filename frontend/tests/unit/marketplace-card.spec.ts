
import Vue from 'vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import MarketplaceCard from '@/components/cards/MarketplaceCard.vue';

import { deleteMarketplaceCard, User } from '@/api/internal';

jest.mock('@/api/internal', () => ({
  deleteMarketplaceCard: jest.fn(),
}));

Vue.use(Vuetify);

const localVue = createLocalVue();

const testUser: User = {
  id: 2,
  firstName: 'test_firstname',
  lastName: 'test_lastname',
  email: 'test_email',
  homeAddress: { country: 'test_country', city: 'test_city', district: 'test_district'},
};

const testMarketplaceCard = {
  id: 1,
  creator: testUser,
  section: 'ForSale',
  created: '2021-03-10',
  title: 'test_card_title',
  description: 'test_card_description',
  keywords: [{id: 3, name: 'test_keyword_1'}, {id: 4, name: 'test_keyword_2'}],
};

describe('MarketplaceCard.vue', () => {
  let appWrapper: Wrapper<any>;
  let wrapper: Wrapper<any>;
  let vuetify: Vuetify;

  /**
   * Finds the required button in the MarketplaceCard card by specifying the
   * button text in the button.
   * @returns A Wrapper around the required button
   */
  function findButton(buttonText: string, specifiedWrapper: Wrapper<any>) {
    const buttons = specifiedWrapper.findAllComponents({ name: 'v-btn' });
    const filtered = buttons.filter(button => button.text().includes(buttonText));
    expect(filtered.length).toBe(1);
    return filtered.at(0);
  }

  /**
   * Finds the delete confirmation dialog box upon clicking the delete button
   * @returns the delete confirmation dialog box
   */
  async function findDeleteConfirmationDialog() {
    const deleteButton = wrapper.findComponent({ ref: 'deleteButton' });
    await deleteButton.trigger('click');

    const dialogs = wrapper.findAllComponents({ name: "v-dialog" });
    return dialogs.at(0)
  }

  beforeEach(() => {
    vuetify = new Vuetify();

    // Creating wrapper around MarketplaceCard with data-app to appease vuetify
    const App = localVue.component('App', {
      components: { MarketplaceCard },
      template: `
      <div data-app>
        <MarketplaceCard :content="testMarketplaceCard"/>
      </div>`
    });

    const elem = document.createElement('div');
    document.body.appendChild(elem);

    appWrapper = mount(App, {
      localVue,
      vuetify,
      attachTo: elem,
      data() {
        return {
          testMarketplaceCard: testMarketplaceCard
        };
      }
    });

    wrapper = appWrapper.getComponent(MarketplaceCard);
  });

  it('Must match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Must contain the creator name", () => {
    expect(wrapper.text()).toContain('By test_firstname test_lastname');
  });

  it("Must contain the creator address with suburb and city if provided", () => {
    expect(wrapper.text()).toContain('From test_district, test_city');
  });

  it("Must contain the creator address with city and country if provided", async () => {
    await wrapper.setData({
      content: {
        creator: {
          homeAddress: {
            district: undefined,
            city: 'test_city',
            country: 'test_country',
          }
        }
      },
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain('From test_city, test_country');
  });

  it("Must contain the creator country if only field provided", async () => {
    await wrapper.setData({
      content: {
        creator: {
          homeAddress: {
            district: undefined,
            city: undefined,
            country: 'test_country',
          }
        }
      },
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain('From test_country');
  });

  it("Must contain the title", () => {
    expect(wrapper.text()).toContain('test_card_title');
  });

  it("Must contain the description", () => {
    expect(wrapper.text()).toContain('test_card_description');
  });

  it("Must contain keyword names", () => {
    expect(wrapper.text()).toContain("test_keyword_1");
    expect(wrapper.text()).toContain("test_keyword_2");
  });

  it("Must contain posted date", () => {
    expect(wrapper.text()).toContain('Posted 10 Mar 2021');
  });

  it("Must trigger delete confirmation dialog box upon clicking delete icon", async () => {
    expect(wrapper.vm.deleteCardDialog).toBeFalsy();
    //This button is an icon, so a reference is used to identify it instead of its button text
    const deleteButton = wrapper.findComponent({ ref: 'deleteButton' });
    await deleteButton.trigger('click');
    expect(wrapper.vm.deleteCardDialog).toBeTruthy();
  });

  it("The deleteMarketplaceCard method must be called and the dialog box should not be visible, upon clicking the delete button in the confirmation dialog box", async () => {
    const deleteConfirmationDialog = await findDeleteConfirmationDialog();
    const dialogDeleteButton = findButton('Delete', deleteConfirmationDialog);
    await dialogDeleteButton.trigger("click");
    expect(deleteMarketplaceCard).toBeCalledWith(testMarketplaceCard.id);
    expect(wrapper.vm.deleteCardDialog).toBeFalsy();
  });

  it("The dialog box should not be visible if the cancel button is clicked in the confirmation dialog box", async () => {
    const deleteConfirmationDialog = await findDeleteConfirmationDialog();
    const dialogCancelButton = findButton('Cancel', deleteConfirmationDialog);
    await dialogCancelButton.trigger("click");
    expect(wrapper.vm.deleteCardDialog).toBeFalsy();
  });

  
});
