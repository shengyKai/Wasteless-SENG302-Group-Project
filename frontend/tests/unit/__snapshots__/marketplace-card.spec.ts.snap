// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketplaceCard.vue Must match snapshot 1`] = `
<div class="d-flex flex-column v-card v-sheet theme--light" style="min-height: 250px;">
  <div class="v-card__title my-n1 title">
    <div class="mr-1">
      test_card_title
    </div>
    <!---->
  </div>
  <div class="v-card__text my-n2 flex-grow-1 d-flex flex-column justify-space-between">
    <div><strong>
        From test_district, test_city
        <br></strong>
      <div class="d-flex justify-space-between flex-wrap"><span class="mr-1"><em>By test_firstname test_lastname</em></span> <span>
          Posted 10 Mar 2021
        </span></div>
      test_card_description
    </div>
    <div class="v-flex mx-n1"><span class="mx-1 my-1 v-chip theme--light v-size--small primary"><span class="v-chip__content">
        test_keyword_1
      </span></span><span class="mx-1 my-1 v-chip theme--light v-size--small primary"><span class="v-chip__content">
        test_keyword_2
      </span></span></div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="v-card__actions">
    <div class="spacer"></div> <button type="button" class="v-icon notranslate mr-2 v-icon--link mdi mdi-pencil theme--light primary--text"></button> <button type="button" class="v-icon notranslate v-icon--link mdi mdi-trash-can theme--light primary--text"></button>
    <div role="dialog" class="v-dialog__container">
      <!---->
    </div>
  </div>
  <!---->
</div>
`;
